### PR TITLE
simplify selector implementation

### PR DESCRIPTION
**Pull request by lesbroot**

### CODE REVIEW

**Comment by kmate** [comment]
So, actually this is a preparation to do #38 the right way - we just figured out that it's near to impossible to do that right without this.

File: core/src/main/scala/ste/selector.scala [hunk 1]
```diff
@@ -89,130 +74,98 @@ object StructTypeSelector extends SelectorImplicits {
...
-      val structs = childPrefixes.map { p =>
-        val cols = fields.filter(_.isChildrenOf(p)).map(f => dfHead(f.quotedString).as(f.getSuffix))
-        struct(cols :_*).as(p.toString)
-      }
-      val dfStruct = dfHead.select((structs ++ restCols) :_*)
-      val nestedCols = getNestedColumns(childPrefixes, dfStruct, fl)
-      orderedSelect(dfStruct, nestedCols, fields)
-    }.getOrElse(dfHead)
-    tSelector.select(dfNested, parentPrefixes, flatten.tail)
+    val hColumn = hSelector.value.select(addPrefix(prefix, fieldName, parentFlatten), flatten.head).as(fieldName)
```
Comment (BenFradet) [inline core/src/main/scala/ste/selector.scala]: can we use headOption, or a NonEmptyList for flatten?

File: core/src/main/scala/ste/selector.scala [hunk 2]
```diff
@@ -89,130 +74,98 @@ object StructTypeSelector extends SelectorImplicits {
...
+      .getOrElse(s.select(prefix, flatten))
   }
 
   implicit def mapSelector[K, V](
     implicit s: DataTypeSelector[V]
-  ): DataTypeSelector[Map[K, V]] = DataTypeSelector.pure { (df, prefixes) =>
-    s.select(df, prefixes)
+  ): DataTypeSelector[Map[K, V]] = DataTypeSelector.pure { (prefix, flatten) =>
+    flatten
+      .map(_.keys.flatMap(k => Seq(lit(k), s.select(addPrefix(prefix, k, flatten), flatten))))
```
Comment (BenFradet) [inline core/src/main/scala/ste/selector.scala]: same with List

File: core/src/main/scala/ste/selector.scala [hunk 3]
```diff
@@ -89,130 +74,98 @@ object StructTypeSelector extends SelectorImplicits {
...
+      .map(map(_: _*))
+      .getOrElse(s.select(prefix, flatten))
   }
 
   implicit class FlattenedDataFrame(df: DataFrame) {
-    def asNested[A : Encoder : StructTypeSelector]: Dataset[A] = selectNested.as[A]
+    def asNested[A <: Product : Encoder : StructTypeSelector]: Dataset[A] = selectNested.as[A]
```
Comment (BenFradet) [inline core/src/main/scala/ste/selector.scala]: why was this needed?

Comment (BenFradet) [reply core/src/main/scala/ste/selector.scala]: for the tuples I'm guessing?

File: core/src/main/scala/ste/selector.scala [hunk 4]
```diff
@@ -89,130 +74,98 @@ object StructTypeSelector extends SelectorImplicits {
...
     }
 }
 
 trait SelectorImplicits {
-  implicit val hnilSelector: AnnotatedStructTypeSelector[HNil] =
-    AnnotatedStructTypeSelector.pure((df, _, _) => df)
+  type Prefix = Vector[String]
+
+  def addPrefix(prefix: Prefix, s: String, flatten: Option[Flatten]): Prefix = flatten match {
+    case Some(_) => prefix.dropRight(1) :+ prefix.lastOption.map(p => s"$p.$s").getOrElse(s)
```
[the remark this answers is not in the document]
Comment (lesbroot) [reply core/src/main/scala/ste/selector.scala]: yes it's safe:
```scala
scala> List().dropRight(1)
res2: List[Nothing] = List()
```

File: core/src/main/scala/ste/selector.scala (hunk 1 above)
Comment (lesbroot) [reply core/src/main/scala/ste/selector.scala]: for `NonEmptyList`, I would need to add `cats` as a dependency, is that all right?

Comment (lesbroot) [reply core/src/main/scala/ste/selector.scala]: nvm, I think that wouldn't fork for the `HNil` case anyway.

File: core/src/main/scala/ste/selector.scala (hunk 3 above)
Comment (lesbroot) [reply core/src/main/scala/ste/selector.scala]: I was just experimenting, it is not needed, I removed it.
